File: boj/BinarySearch/BOJ_2539.swift
/*
 백준 2539 모자이크
 22.06.08
 https://github.com/skyqnaqna/algorithm_study
 */

import Foundation

final class FileIO {
  private var buffer:[UInt8]
  private var index: Int
  
  init(fileHandle: FileHandle = FileHandle.standardInput) {
    buffer = Array(fileHandle.readDataToEndOfFile())+[UInt8(0)] // 인덱스 범위 넘어가는 것 방지
    index = 0
  }
  
  @inline(__always) private func read() -> UInt8 {
    defer { index += 1 }
    
    return buffer.withUnsafeBufferPointer { $0[index] }
  }
  
  @inline(__always) func readInt() -> Int {
    var sum = 0
    var now = read()
    var isPositive = true
    
    while now == 10
            || now == 32 { now = read() } // 공백과 줄바꿈 무시
    if now == 45{ isPositive.toggle(); now = read() } // 음수 처리
    while now >= 48, now <= 57 {
      sum = sum * 10 + Int(now-48)
      now = read()
    }
    
    return sum * (isPositive ? 1:-1)
  }
  
  @inline(__always) func readString() -> String {
    var str = ""
    var now = read()
    
    while now == 10
            || now == 32 { now = read() } // 공백과 줄바꿈 무시
    while now != 10
            && now != 32 && now != 0 {
      str += String(bytes: [now], encoding: .ascii)!
      now = read()
    }
    
    return str
  }
}

let io = FileIO()

let n = io.readInt(), m = io.readInt()

var target = [(Int, Int)]()
var l = 1, r = max(n, m) + 1

let k = io.readInt(), cnt = io.readInt()

for _ in 0 ..< cnt {
  let row = io.readInt(), col = io.readInt()
  target.append((row, col))
  
  if l < row {
    l = row
  }
}

target.sort(by: { $0.1 > $1.1 })

func getCnt(_ size: Int) -> Int {
  var res = 0
  var temp = target
  
  while !temp.isEmpty {
    let (r, c) = temp.removeLast()
    
    if r > size { return 101 }
    
    while !temp.isEmpty && temp.last!.0 <= size && temp.last!.1 < c + size {
      temp.removeLast()
    }
    
    res += 1
  }
  
  return temp.isEmpty ? res : 101
}

var answer = 0

while l <= r {
  let mid = (l + r) / 2
  let res = getCnt(mid)
  
  if res <= k {
    r = mid - 1
    answer = mid
  } else {
    l = mid + 1
  }
}

print(answer)

File: boj/BinarySearch/BOJ_2539.cpp
/*
 백준 2539 모자이크
 22.06.08
 https://github.com/skyqnaqna/algorithm_study
 */

#include <bits/stdc++.h>

using namespace std;
using ll = long long;
using dd = double;
using pii = pair<int, int>;
using pll = pair<ll, ll>;
using pil = pair<int, ll>;

#define INF (int)2e9
#define endl "\n"

void input() {
  freopen("input.txt", "r", stdin);
}

void output() {
  ifstream readFile;
  readFile.open("output.txt");

  if (readFile.is_open()) {
    while (!readFile.eof()) {
      string a;
      getline(readFile, a);
      //            ans.push_back(stoi(a));
    }
    readFile.close();
  }
}

int dy[] = {-1, 0, 1, 0, -1, -1, 1, 1};
int dx[] = {0, 1, 0, -1, -1, 1, -1, 1};

//=======================================//

int n, m, k, cnt;
vector <pii> target;

bool cmp(pii a, pii b) {
  return a.second < b.second;
}

int getCnt(int size) {
  int res = 0, idx = 0;
  
  while (idx < cnt) {
    int r = target[idx].first;
    int c = target[idx++].second;
    
    if (r > size) return 101;
    
    while (idx < cnt && target[idx].first <= size && target[idx].second < c + size) {
      ++idx;
    }
    
    ++res;
  }
  
  return idx < cnt ? 101 : res;
}

int main() {
  ios::sync_with_stdio(0); cin.tie(0);
  
  cin >> n >> m >> k >> cnt;
  
  int row, col, l = 1, r = max(n, m);
  for (int i = 0; i < cnt; ++i) {
    cin >> row >> col;
    target.push_back({row, col});
    
    if (l < row) l = row;
  }
  
  sort(target.begin(), target.end(), cmp);
  
  
  int mid, answer = 0;
  while (l <= r) {
    mid = (l + r) / 2;
    int res = getCnt(mid);
    
    if (res <= k) {
      r = mid - 1;
      answer = mid;
    } else l = mid + 1;
  }
  
  cout << answer << endl;
  
  return 0;
}
